lib/util.js

```diff
@@ -361,262 +371,229 @@ function formatValue(ctx, value, recurseTimes) {
 
   // Provide a hook for user-specified inspect functions.
   // Check that value is an object with an inspect function on it
-  if (ctx.customInspect && value) {
-    const maybeCustomInspect = value[customInspectSymbol] || value.inspect;
-
-    if (typeof maybeCustomInspect === 'function' &&
-        // Filter out the util module, its inspect function is special
-        maybeCustomInspect !== exports.inspect &&
-        // Also filter out any prototype objects using the circular check.
-        !(value.constructor && value.constructor.prototype === value)) {
-      let ret = maybeCustomInspect.call(value, recurseTimes, ctx);
-
-      // If the custom inspection method returned `this`, don't go into
-      // infinite recursion.
-      if (ret !== value) {
-        if (typeof ret !== 'string') {
-          ret = formatValue(ctx, ret, recurseTimes);
-        }
-        return ret;
+  const customInspect = isCustomInspect(ctx, value);
+  if (customInspect !== undefined) {
+    if (ctx.seen === undefined) {
+      ctx.seen = [];
+      // A customInspect could be set only for partial objects.
+      // To prevent a breaking change, fill the lazy added array with all
+      // gathered objects.
+      if (ctx.seenSet !== undefined)
+        ctx.seenSet.forEach((val) => ctx.seen.push(val));
+    }
+    let ret = customInspect.call(value, recurseTimes, ctx);
+    // If the custom inspection method returned `this`, don't go into
+    // infinite recursion.
+    if (ret !== value) {
+      if (typeof ret !== 'string') {
+        ret = formatValue(ctx, ret, recurseTimes);
       }
+      return ret;
     }
   }
 
+  var type = typeof value;
```

Comment:
Hm, I think it is nicer not to write `typeof x` many times over and `type` is currently redefined in [line 531](https://github.com/nodejs/node/pull/14881/files#diff-3deb3f32958bb937ae05c6f3e4abbdf5R531). But I can change it if you want me to.